snuba/query/joins/classifier.py

```diff
@@ -0,0 +1,250 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, replace
+from functools import partial
+from typing import (
+    Callable,
+    List,
+    Mapping,
+    MutableMapping,
+    NoReturn,
+    Optional,
+    Sequence,
+    Set,
+)
+
+from snuba.query.expressions import (
+    Argument,
+    Column,
+    CurriedFunctionCall,
+    Expression,
+    ExpressionVisitor,
+    FunctionCall,
+    Lambda,
+    Literal,
+    SubscriptableReference,
+)
+
+
+def assert_never(value: NoReturn) -> NoReturn:
+    assert False, f"Unhandled value: {value} ({type(value).__name__})"
+
+
+@dataclass(frozen=True)
+class SubExpression(ABC):
+    """
+    Data structure maintained when visiting an Expression in the query.
+    This keeps track of the main branch (the node we are visiting) as
+    well as all the branches we cut that were under the main expression.
+    Those cut branches should be pushed down into subqueries.
+    """
+
+    main_expression: Expression
+
+    @abstractmethod
+    def cut_branch(self) -> MainQueryExpression:
+        """
+        Cut the expression tree dividing the subtree/s that need to be
+        pushed down into the subqueries from the expression that must
+        remain in the main query and that references the subtrees.
+        """
+
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class MainQueryExpression(SubExpression):
+    """
+    A SubExpression that contains subtrees to be pushed down and a main
+    expression for the main query.
+    """
+
+    cut_branches: Mapping[str, Set[Expression]]
+
+    def cut_branch(self) -> MainQueryExpression:
+        return self
+
+
+@dataclass(frozen=True)
+class SubqueryExpression(SubExpression):
+    """
+    A SubExpression that can be entirely pushed into a subquery since
+    it references columns from one subquery only.
+    """
+
+    subquery_alias: str
+
+    def cut_branch(self) -> MainQueryExpression:
+        """
+        This Returns a simple column that references the expression in
+        the subquery and cut the entire expression tree.
+        """
+        assert (
+            self.main_expression.alias
+        ), f"Invalid expression in join query {self.main_expression}. Missing alias"
+        return MainQueryExpression(
+            main_expression=Column(
```

Comment:
I think the Column name is the misleading one. What we reference as column in the ast is really a reference to something that has a name in the abstract data model the query reasons onto. 
In the clickhouse query, these are physical clickhouse columns, because the data model of those queries are clickhouse tables.
In the logical query, these are reference to any logical column in the from clause (the data source in the from clause always has a ColumnSet as schema, whether those are real columns or not).
In the case of subqueries. the main query reference expressions exposed in the subquery, whether these are columns or complex expressions, the ColumnSet in the data source eposes those names.

So in theory I would say we misuse the `column` name. In practice we could add a separate concept, but for all practical purposes Column and Reference would serve the exact same purpose and work the same way (same fields, same usage).
(This is true because in no place in query processing we ever need to make a distinction between what is a real clickhouse column and what is a reference to another expression, should that be needed one day we should probably split these two concepts, but in that case I would add a new PhysicalColumn instead of a Adding a Reference concept as the Column today is always a reference).

We could do that, though I would definitively keep it a separate effort since that implies touching all visitors and all places where we delas with columns.

Reply:
Thinking over that, we should never have in the same query references to expressions and references to physical columns.
- Logical simple query: being all aliases unpacked, all Column objects are reference to a "column" in the abstarct data model, which may not be a column
- Logical composite query: columns in the subqueries are like above, columns in the parent query MUST all be references as we do not allow a mix of entities and subqueries in the FROM clause at this level.
- Clickhouse simple query: all columns are reference to something exposed by the clickhouse schema of the table/view, which is always a column whether real or materialized, but we do not make a difference. Aliases are all unpacked
- Clickhouse composite query. Like the logical query.

The only exception could be if we unpack a subquery right before formatting for optimization reasons (but after that there cannot be further processing).
This makes me think that having two separate concepts in the AST could actually be counter productive as they should never be together in the same query.

(if we change some of the assumptions here, the whole query processing changes in ways that are more radical than this decision)

Reply:
The Reference I had in mind would have some differences to Column (aka "Physical column"). Column has 3 fields - alias, optional table name and column name, where the Reference would only have 2 fields - table name (which would become mandatory) and the alias.

Reply:
I think you would always need three:
1) the entity/table you are referring to: table_name
2) the name of the thing you are referring to: column_name in case of columns 
3) its own alias if that reference has to be referred to somewhere else: alias. Imagine you have a three level nested queries. The inner one has columns, the one in the middle has references to those columns, the outer one has references to the references of the middle one. So references must be referrable (?) as well.